### PR TITLE
5.01 AST support #2 - Arcana Logs with gauge states

### PR DESCRIPTION
**Pull request by suShirow**
Upgrade to TS
Fix incorrect data IDs
Delete unused code
Please Don't Hurt Me I'm Just A Squishy Card Vendo

### CODE REVIEW

**Comment by ackwell** [comment]
Needs some fixes to appease the linting gods

**Comment by suShirow** [comment]
My prayers have been answered, 
      thank you for guiding my code true, 
O all-seeing linter;  //

File: src/parser/jobs/ast/modules/ArcanaTracking.tsx [hunk 1]
```diff
@@ -0,0 +1,509 @@
...
+		const sleeveDrawLog: CastEvent[] = []
+
+		for (const event of events) {
+			if ((event.timestamp > startTime && !prepullSleeve)
+				|| event.timestamp - startTime >= (STATUSES.SLEEVE_DRAW.duration * 1000)
+				) {
+				// End loop if: 1. Max duration of sleeve draw status passed 2. No prepull sleeve detected
+				break
+
+				} else if (!prepullSleeve && event.type === 'applybuff' && STATUSES.SLEEVE_DRAW.id === event.ability.guid
```
Comment (ackwell) [inline src/parser/jobs/ast/modules/ArcanaTracking.tsx]: looks like indentation is ded. linter isn't complaining, though, so not toooo worried.

File: src/parser/jobs/ast/modules/ArcanaTracking.tsx [hunk 2]
```diff
@@ -0,0 +1,509 @@
...
+
+		if (sleeveDrawLog.length > 0 && sleeveDrawLog.length <= 2) {
+			this.cardStateLog[0].sleeveState = sleeveDrawLog.length
+		}
+
+		return events
+
+	}
+
+	public get getCardLogs() {
```
Comment (ackwell) [inline src/parser/jobs/ast/modules/ArcanaTracking.tsx]: not sure i'm following what's going on here - `cardStateLog` should be `[]` if `length === 0` anyway, and it's initiated with an initial entry.

---

also avoid naming getters with `get...`, it looks kinda weird at call sites

File: src/parser/jobs/ast/modules/ArcanaTracking.tsx [hunk 3]
```diff
@@ -0,0 +1,509 @@
...
+	 * Determine exactly when they casted their prepull arcana
+	 * If they had overwritten this buff, it will falsly pull back the timestamp of their prepull cast, but since we are guessing, it may as well be the same.
+	 */
+	private offPrepullArcana(event: BuffEvent) {
+		if (event.timestamp >= this.parser.fight.start_time + (STATUSES.THE_BALANCE.duration * 1000)) {
+			return
+		}
+
+		this.prepullArcanas.forEach(arcanaBuff => {
+			if (arcanaBuff.ability.guid === event.ability.guid
```
Comment (ackwell) [inline src/parser/jobs/ast/modules/ArcanaTracking.tsx]: if you invert this `if` and the one below it, you can `return` in them to shortcut out without needing to indent

File: src/parser/jobs/ast/modules/ArcanaTracking.tsx [hunk 4]
```diff
@@ -0,0 +1,509 @@
...
+					}
+					cardStateItem.lastEvent = {...arcanaCastEvent}
+					cardStateItem.drawState = DrawnType.NOTHING
+					cardStateItem.sealState = CLEAN_SEAL_STATE
+
+					this.cardStateLog.unshift(cardStateItem)
+					this.pullIndex++
+				}
+
+				return
```
Comment (ackwell) [inline src/parser/jobs/ast/modules/ArcanaTracking.tsx]: doesn't look like this is doing much

File: src/parser/jobs/ast/modules/ArcanaTracking.tsx [hunk 5]
```diff
@@ -0,0 +1,509 @@
...
+	 *
+	 * a) If it can't find any clear reason why the player had lost the buff, let's do a retconsearch to figure out since when they had it
+	 *
+	 * b) If they lost the buff with no link to any timestamp, it could be a /statusoff macro.
+	 *    Creates a new entry as this is technically also a card action.
+	 *
+	 */
+	private offDrawnStatus(event: BuffEvent) {
+
+		if (DRAWN_ARCANA.includes(event.ability.guid)) {
```
Comment (ackwell) [inline src/parser/jobs/ast/modules/ArcanaTracking.tsx]: same story w/ `if`

---

also, `isPaired` can be done with `.some` :meowmorning:

File: src/parser/jobs/ast/modules/ArcanaTracking.tsx [hunk 6]
```diff
@@ -0,0 +1,509 @@
...
+				if (cardState.lastEvent.type === 'pull') {
+					this.cardStateLog[index].drawState = drawnStatus ? drawnStatus : DrawnType.NOTHING
+					return this.pullStateInitialized = true
+				}
+			})
+		}
+	}
+
+	private onComplete() {
+		// Nothing to suggest yet
```
Comment (ackwell) [inline src/parser/jobs/ast/modules/ArcanaTracking.tsx]: probably can drop the hook then

File: src/parser/jobs/ast/modules/ArcanaTracking.tsx (hunk 2 above)
Comment (suShirow) [reply src/parser/jobs/ast/modules/ArcanaTracking.tsx]: Okay, this one's my bad - this is leftover code from the 4.0 impl of this module, which did have a use for that check. I'll just make it a simple return getter.

File: src/parser/jobs/ast/modules/ArcanaTracking.tsx (hunk 3 above)
Comment (suShirow) [reply src/parser/jobs/ast/modules/ArcanaTracking.tsx]: I need to master this technique :woah:

**Comment by suShirow** [comment]
@ackwell code purification rites complete

**Comment by suShirow** [comment]
There is a known bug right now that can be seen here (in my own log!)
/analyse/baWK4AxRfHqPBvt2/3/1/
Basically if they have only 1 stack of sleeve draw going into the fight, it will detect none. Going to try to toss the fix into this PR tonight.

File: src/parser/jobs/ast/modules/ArcanaTracking/ArcanaSuggestions.module.css [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+.buffIcon {
+	width: 1.4285em;
+	margin-left: .25em;
+	vertical-align: bottom;
+}
+
+.buffPlaceholder, .buffUnknown {
+	width: 1.4285em;
+    margin: 0 0 0 0.25em;
+    vertical-align: bottom;
```
Comment (Fank) [inline src/parser/jobs/ast/modules/ArcanaTracking/ArcanaSuggestions.module.css]: wrong indent

File: src/parser/jobs/ast/modules/ArcanaTracking/ArcanaSuggestions.module.css [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+	position: relative;
+}
+
+.sealIconContainer {
+    display: inline-block;
+	position: relative;
+	width: 65px;
+	height: 26px;
+    vertical-align: bottom;
+    background-image: url('./seals_jobhud_simple.png')
```
Comment (Fank) [inline src/parser/jobs/ast/modules/ArcanaTracking/ArcanaSuggestions.module.css]: same here

**Comment by suShirow** [comment]
Fixed issue with a less than savory normaliser change, but it works and it'll do for now and hopefully for the start of savage :blobsweat:

File: src/parser/jobs/ast/modules/ArcanaTracking/ArcanaSuggestions.module.css (hunk 7 above)
Comment (suShirow) [reply src/parser/jobs/ast/modules/ArcanaTracking/ArcanaSuggestions.module.css]: Thank you! My VScode betrayed me

File: src/parser/jobs/ast/modules/ArcanaTracking/ArcanaTracking.tsx [hunk 9]
```diff
@@ -110,27 +110,25 @@ export default class ArcanaTracking extends Module {
...
 			if ((event.timestamp > startTime && !prepullSleeve)
 				|| event.timestamp - startTime >= (STATUSES.SLEEVE_DRAW.duration * 1000)
 				) {
-				// End loop if: 1. Max duration of sleeve draw status passed 2. No prepull sleeve detected
-				break
+					// End loop if: 1. Max duration of sleeve draw status passed 2. No prepull sleeve detected
+					break
 
-				} else if (!prepullSleeve && event.type === 'applybuff' && STATUSES.SLEEVE_DRAW.id === event.ability.guid
+				} else if (!prepullSleeve && event.type === 'applybuff' && this.isBuffEvent(event) && STATUSES.SLEEVE_DRAW.id === event.ability.guid
```
Comment (ackwell) [inline src/parser/jobs/ast/modules/ArcanaTracking/ArcanaTracking.tsx]: nit; you can include the `event.type` check in the type guard, so you don't need to include both in the conditional

File: src/parser/jobs/ast/modules/ArcanaTracking/ArcanaTracking.tsx [hunk 10]
```diff
@@ -181,45 +176,43 @@ export default class ArcanaTracking extends Module {
...
+					abilityIcon: _.replace(_.replace(arcanaAction.icon, 'https://xivapi.com/i/', ''), '/', '-'),
+				},
+				timestamp: event.timestamp - (STATUSES.THE_BALANCE.duration * 1000),
+				type: 'cast',
+				sourceIsFriendly: true,
+				targetIsFriendly: true,
+				sourceID: event.sourceID,
+				targetID: event.targetID,
 			}
+			cardStateItem.lastEvent = {...arcanaCastEvent}
```
Comment (ackwell) [inline src/parser/jobs/ast/modules/ArcanaTracking/ArcanaTracking.tsx]: nit; spreading is mainly best for when you want a copy of the object/array so you don't/can't modify the original - given the original in this case is being constructed right above, that's probably not a concern.

**Pull request merged**